cgrep.py

```diff
@@ -0,0 +1,443 @@
+# Copyright 2011 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# Simple util to grep through network and service definitions.
+# Examples:
+#   To find out which tokens contain "10.4.3.1" use
+#   $ cgrep.py -i 10.4.3.1
+#
+#   To find out if token 'FOO' includes ip "1.2.3.4" use
+#   $ cgrep.py -t FOO -i 1.2.3.4
+#
+#   To find the difference and union of tokens 'FOO' and 'BAR' use
+#   $ cgrep.py -c FOO BAR
+#
+#   To find the difference of network tokens to which 2 IPs belong use
+#   $ cgrep.py -g 1.1.1.1 2.2.2.2
+#
+#   To find which IPs are in the 'FOO' network token use
+#   $ cgrep.py -o FOO
+#
+#   To find which port & protocol pairs are in a service token 'FOO' use
+#   $ cgrep.py -s FOO
+#
+#   To find which service tokens contain port '22' and protocol 'tcp' use
+#   $ cgrep.py -p 22 tcp
+#
+
+import argparse
+import pprint
+from lib import naming
+from lib import nacaddr
+
+
+__author__ = ("watson@google.com (Tony Watson)", "jhrbek@fb.com (Joe Hrbek)"
+              "rtierney@fb.com (Ryan Tierney)")
+
+
+def is_valid_ip(arg):
+  """ Validates a value to be an IP or not.
+
+  :arg str: the potential IP address
+  :returns str: the IP they gave us (if it's an ip, else raises an error)
+  """
+  try:
+    nacaddr.IP(arg)
+  except:
+    raise argparse.ArgumentTypeError("%s is an invalid ip address" % arg)
+  return arg
+
+
+def cli_options():
+  """ builds the argparse options for cgrep.
+
+  :returns argparse.ArgumentParser: the arguments, ready to be parsed.
+  """
+
+  parser = argparse.ArgumentParser(
+      description="c[apirca]grep",
+      formatter_class=argparse.RawTextHelpFormatter
+  )
+
+  parser.add_argument("-d", "--def", dest="defs",
+                      help="Network Definitions directory location. \n",
+                      default="./def")
+
+  # -i and -t can be used together, but not with any other option.
+  ip_group = parser.add_argument_group()
+  # take 1 or more IPs
+  ip_group.add_argument("-i", "--ip", dest="ip", nargs='+', type=is_valid_ip,
```

Comment:
You have some inconsistant quote use here and throughout the file. Please fix these.